### PR TITLE
fix(core): Order of module destroy should be the reverse of module init

### PR DESCRIPTION
**Pull request by wenlong-chen**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Suppose A depends on B, currently, A.onModuleDestroy is executed after B.onModuleDestroy. 

Issue Number: https://github.com/nestjs/nest/issues/14118

## What is the new behavior?

Suppose A depends on B, A.onModuleDestroy is executed before B.onModuleDestroy. 

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
Let's consider the following example:

1. A depends on B (AModule -> BModule)
2. Before tearing down, A needs to perform some actions and it needs B (its dependency) to do this
3. This is currently feasible as the order reflects relations between providers (a->b), meaning, the outer gets executed first
4. If we merged this PR, this would no longer be doable as B would have been destroyed before A (meaning, when `onModuleDestroy` was called for A, A would no longer be able to interact with any of its dependencies coming from B)

In summary, although reversing the order might seem intuitive, it actually creates more issues than it resolves. Especially since B should never interact directly with A unless a circular relationship is introduced—in which case, the order wouldn’t matter anyway.

**Comment by wenlong-chen** [comment]
@kamilmysliwiec I’ve created a minimal reproduction code. Please take a look. https://github.com/wenlong-chen/nest-on-module-destory-bug

**Comment by kamilmysliwiec** [comment]
LGTM
